tiatoolbox/models/controller/semantic_segmentor.py

```diff
@@ -0,0 +1,1017 @@
+        scale_factor = scale_factor[0]  # what if ratio x != y
+
+        def sel_func(coord: np.ndarray):
+            """Accept coord as long as its box contains bits of mask."""
+            coord_in_real_mask = np.ceil(scale_factor * coord).astype(np.int32)
+            tl_x, tl_y, br_x, br_y = coord_in_real_mask
+            roi = mask_reader.img[tl_y:br_y, tl_x:br_x]
+            return np.sum(roi > 0) > 0
+
+        flags = [sel_func(bound) for bound in bounds]
+        return np.array(flags)
+
+    @staticmethod
+    def get_reader(img_path: str, mask_path: str, mode: str, auto_get_mask: bool):
+        """Define how to get reader for mask and source image."""
+        img_path = pathlib.Path(img_path)
+        reader = get_wsireader(img_path)
+
+        mask_reader = None
+        if mask_path is not None:
+            if not os.path.isfile(mask_path):
+                raise ValueError("`mask_path` must be a valid file path.")
+            mask = imread(mask_path)  # assume to be gray
+            mask = cv2.cvtColor(mask, cv2.COLOR_RGB2GRAY)
+            mask = np.array(mask > 0, dtype=np.uint8)
+
+            mask_reader = VirtualWSIReader(mask)
+            mask_reader.info = reader.info
+        elif auto_get_mask and mode == "wsi" and mask_path is None:
+            # if no mask provided and `wsi` mode, generate basic tissue
+            # mask on the fly
+            mask_reader = reader.tissue_mask(resolution=1.25, units="power")
+            mask_reader.info = reader.info
+        return reader, mask_reader
+
+    def _predict_one_wsi(
+        self,
+        wsi_idx: int,
+        ioconfig: IOSegmentorConfig,
+        save_path: str,
+        mode: str,
+    ):
+        """Make a prediction on tile/wsi.
+
+        Args:
+            wsi_idx (int): index of the tile/wsi to be processed within `self`.
+            ioconfig (:class:`IOSegmentorConfig`): object which defines I/O
+              placement during inference and when assembling back to full tile/wsi.
+            loader (torch.Dataloader): loader object which return batch of data
+              to be input to model.
+            save_path (str): location to save output prediction as well as possible
+              intermediat results.
+            mode (str): `tile` or `wsi` to indicate run mode.
+
+        """
+        wsi_path = self.imgs[wsi_idx]
+        mask_path = None if self.masks is None else self.masks[wsi_idx]
+        wsi_reader, mask_reader = self.get_reader(
+            wsi_path, mask_path, mode, self.auto_generate_mask
+        )
+
+        resolution = ioconfig.highest_input_resolution
+        if (
+            isinstance(wsi_reader, VirtualWSIReader)
+            and resolution["units"] != "baseline"
+        ):
+            raise ValueError("Inference on `tile` only use `units='baseline'` !")
+        wsi_proc_shape = wsi_reader.slide_dimensions(**resolution)
+
+        # * retrieve patch and tile placement
```

Comment:
What is the `*` for at the start of the comment. I have see this a few times but it is a mystery to me.